harness/tests/experiment/tensorflow/test_estimator_trial.py

```diff
@@ -197,6 +198,42 @@ def make_trial_controller_fn(
 
         utils.optimizer_state_test(make_trial_controller_fn, tmp_path)
 
+    def test_callbacks(self) -> None:
+        with tempfile.TemporaryDirectory() as temp_directory:
+            batches_per_step = 5
+            steps = 10
+            validation_freq = 5
+
+            def make_workloads() -> workload.Stream:
+                trainer = utils.TrainAndValidate()
+
+                yield from trainer.send(
+                    steps=steps, validation_freq=validation_freq, batches_per_step=batches_per_step
+                )
+                yield workload.terminate_workload(), [], workload.ignore_workload_response
+
+            hparams = self.hparams.copy()
+            hparams["training_log_path"] = os.path.join(temp_directory, "training.log")
+            hparams["val_log_path"] = os.path.join(temp_directory, "val.log")
+
+            controller = utils.make_trial_controller_from_trial_implementation(
+                trial_class=estimator_xor_model.XORTrialWithCallbacks,
+                hparams=hparams,
+                workloads=make_workloads(),
+                batches_per_step=batches_per_step,
+            )
+            controller.run()
+
+            with open(hparams["training_log_path"], "r") as fp:
+                assert int(fp.readline()) == batches_per_step * steps
+
+            with open(hparams["val_log_path"], "r") as fp:
+                # Because we do not specify number of steps in validation, the first
+                # time we run validation, after going through all the examples, eval
+                # will call `before_run()` then hit an I/O exception. Thus we need
+                # to +1 to the expected number of calls.
```

Comment:
blocking: i don't really understand this comment. if we hit an I/O exception on `before_run()` why doesn't this test fail? might want to add some more to this comment to help future maintainability, or just reframe the test in a way that it doesn't have this edge case.

Reply:
The I/O exception is used by Estimators to determined the end of the dataset on the first epoch if `max_steps` isn't specified (Estimators catches and handles it). For callbacks this means that on the last batch of the first epoch `before_run()` will be called one more than the number of data items present in the dataset, so we have to do + 1 on the expected result.

I simplified the test by moving the `self_idx += 1` to `after_run()` which is not called an extra time. We lose a bit of coverage on `before_run()` but should be ok.